### ISSUE TITLE
3 tests failed with System.Runtime.InteropServices.COMException in x64_release_windows_nt_ilrt

### ISSUE THREAD

**Issue by jiangzeng** [dotnet/coreclr#20801]
**Jobs:**
[https://ci.dot.net/job/dotnet_coreclr/job/master/view/x64/job/x64_release_windows_nt_ilrt/671/](https://ci.dot.net/job/dotnet_coreclr/job/master/view/x64/job/x64_release_windows_nt_ilrt/671/)

**Affected tests:**
```
Interop_COM._NETClients_Primitives_NETClientPrimitives_NETClientPrimitives_._NETClients_Primitives_NETClientPrimitives_NETClientPrimitives_cmd
Interop_COM._NativeClients_Primitives_Primitives_._NativeClients_Primitives_Primitives_cmd
Interop_COM._NETClients_IDispatch_NETClientIDispatch_NETClientIDispatch_._NETClients_IDispatch_NETClientIDispatch_NETClientIDispatch_cmd
```

**Log:**
```
Test Failure: System.Runtime.InteropServices.COMException (0x80040154): Retrieving the COM class factory for component with CLSID {53169A33-E85D-4E3C-B668-24E438D0929B} failed due to the following error: 80040154 Class not registered (Exception from HRESULT: 0x80040154 (REGDB_E_CLASSNOTREG)).
           at NetClient.NumericTests..ctor(Int32 seed) in D:\j\workspace\x64_release_w---21e2e994\tests\src\Interop\COM\NETClients\Primitives\NumericTests.cs:line 22
           at NetClient.Program.Main(String[] doNotUse) in D:\j\workspace\x64_release_w---21e2e994\tests\src\Interop\COM\NETClients\Primitives\Program.cs:line 21
        END EXECUTION OF IL{D}ASM BINARY - FAILED 101 vs 100
        FAILED
        Test Harness Exitcode is : 1

```

**Comment by jiangzeng** [comment]
Similar to issue #20601

**Comment by BruceForstall** [comment]
@AaronRobinsonMSFT

**Comment by AaronRobinsonMSFT** [comment]
This is unrelated to #20601. This is _not_ the test itself, but some kind of ILDASM/ILASM process step. The test itself works as expected. I am not sure what process is going on here.

>BEGIN EXECUTION
06:53:33         D:\j\workspace\x64_release_w---21e2e994\bin\tests\Windows_NT.x64.Release\Tests\Core_Root\ildasm.exe /raweh /out=NETClientAggregation.dasm.il NETClientAggregation.exe
06:53:33         D:\j\workspace\x64_release_w---21e2e994\bin\tests\Windows_NT.x64.Release\Tests\Core_Root\ilasm.exe /output=NETClientAggregation.asm.exe  NETClientAggregation.dasm.il

Source: https://ci.dot.net/job/dotnet_coreclr/job/master/view/x64/job/x64_release_windows_nt_ilrt/671/consoleFull#-5725545192d31e50d-1517-49fc-92b3-2ca637122019

@BruceForstall or @RussKeldorph is there some JIT testing of ILDASM that is being performed here? Do you know what this process is?

**Comment by AaronRobinsonMSFT** [comment]
Okay. This is something to do with the `RunningIlasmRoundTrip` scenario. Anyone have any insight into what this is supposed to do and why it breaks the COM tests? Perhaps a bug in ILDASM?

@jeffschwMSFT

**Comment by RussKeldorph** [comment]
@AaronRobinsonMSFT The test is trying to verify that assemblies can "round-trip" through ildasm->ilasm.  A couple of the failures are known (#19432) but the interop ones are apparently new.  I haven't looked at this one yet, but it's usually pretty easy to figure out why the test is failing from the dasm output.

**Comment by AaronRobinsonMSFT** [comment]
This is actually a relatively straight forward problem. The issue here is that the ILDASM/ILASM round trip tests don't reassemble using the resulting `.res` file from the ILDASM process. This is needed because the COM tests use RegFree COM which requires an [App manifest](https://docs.microsoft.com/en-us/windows/desktop/sbscs/application-manifests). Since the manifest is not present in the newly assembled assembly there is no COM registration and as such the resulting test fails.

As an example the above command should add the `/RESOURCE:NETClientAggregation.dasm.res` flag and argument. However it was observed when trying this locally that ILASM asserts when provided the '/RESOURCE` flag. The assert is at the following:

https://github.com/dotnet/coreclr/blob/2c2cab344c4909da0bd0ebcd6a80c6c16c2598dc/src/dlls/mscorpe/ceefilegenwriter.cpp#L953

From the looks of the code, this should 'just work' since most of the PE resource logic appears to be behind the `FEATURE_PAL` feature.

Based on my quick investigation all that would need to happen form the testing infrastructure side is for the following generated code to be updated to pass the `/RESOURCE` flag in Windows scenarios.

https://github.com/dotnet/coreclr/blob/2c2cab344c4909da0bd0ebcd6a80c6c16c2598dc/tests/src/CLRTest.Jit.targets#L99

Alternatively tests could set the `_IlasmSwitches`, although I am less confident in the `_IlasmSwitches` approach but it is an option.

@jkoritzinsky This might be an interesting test for you to explore how RegFree COM works and issues with debugging it.

**Comment by jkoritzinsky** [comment]
As we spoke offline, resource embedding in ilasm is going to be more difficult than we initially expected. See #20819 for more information.

**Comment by jeffschwMSFT** [comment]
Fixed with #20822
